tools/vhost-user-video/virtio_video_udmabuf.c

```diff
@@ -0,0 +1,175 @@
+// SPDX-License-Identifier: GPL-2.0+
+/*
+ * Virtio Video Device
+ *
+ * Copyright Red Hat, Inc. 2023
+ *
+ * Authors:
+ *     Albert Esteve <aesteve@redhat.com>
+ *
+ * This work is licensed under the terms of the GNU GPL, version 2 or later.
+ * See the COPYING file in the top-level directory.
+ */
+
+#include "qemu/osdep.h"
+#include <sys/ioctl.h>
+#include <sys/mman.h>
+
+#include "vuvideo.h"
+
+struct udmabuf_create {
+    uint32_t memfd;
+    uint32_t flags;
+    uint64_t offset;
+    uint64_t size;
+};
+
+#define UDMABUF_CREATE _IOW('u', 0x42, struct udmabuf_create)
+
+static size_t
+udmabuf_get_size(struct VuVideoDMABuf *buf)
+{
+    return ROUND_UP(buf->length, qemu_real_host_page_size());
+}
+
+static bool
+udmabuf_alloc_bm(struct VuVideoDMABuf *buf)
+{
+    int ret;
+
+    buf->memfd = memfd_create("udmabuf-video-bm", MFD_ALLOW_SEALING);
+    if (buf->memfd < 0) {
+        return false;
+    }
+
+    ret = ftruncate(buf->memfd, udmabuf_get_size(buf));
+    if (ret < 0) {
+        close(buf->memfd);
+        return false;
+    }
+
+    ret = fcntl(buf->memfd, F_ADD_SEALS, F_SEAL_SHRINK);
+    if (ret < 0) {
+        close(buf->memfd);
+        return false;
+    }
+
+    return true;
+}
+
+static void
+udmabuf_free_bm(struct VuVideoDMABuf *buf)
+{
+    close(buf->memfd);
+}
+
+static bool
+udmabuf_map_bm(struct VuVideoDMABuf *buf)
+{
+    g_debug("Map the buffer memory.");
+    buf->start = mmap(NULL, udmabuf_get_size(buf),
+                      PROT_READ | PROT_WRITE, MAP_SHARED, buf->memfd, 0);
+    if (buf->start == MAP_FAILED) {
+        return false;
+    }
+
+    return true;
+}
+
+static void
+udmabuf_unmap_bm(struct VuVideoDMABuf *buf)
+{
+    g_debug("Unmap the buffer memory.");
+    munmap(buf->start, udmabuf_get_size(buf));
+}
+
+static bool
+udmabuf_get_fd(struct VuVideoDMABuf *buf)
+{
+    struct udmabuf_create create = {
+        .memfd = buf->memfd,
+        .offset = 0,
+        .size = udmabuf_get_size(buf),
+    };
+
+    buf->dmafd = ioctl(buf->dev->fd, UDMABUF_CREATE, &create);
+
+    return buf->dmafd >= 0;
```

Comment:
shouldn't you return the file descriptor here instead of a boolean? I'm not familiar with the QEMU device API but that's what I would had expected from a `.get_fd` callback handler.

Reply:
As I was storing the descriptor in the buffer to be able to reuse it, I didn't need to have the caller to assign it, just use it. But I see your point. I will change it so it returns an `int` with the descriptor.